### PR TITLE
Initializer: add InitKind discriminant

### PR DESCRIPTION
**Pull request by WalterBright**
based on a suggestion by @ibuclaw https://github.com/dlang/dmd/pull/8659#issuecomment-425636991

### CODE REVIEW

**Review by PetarKirov — APPROVED** [review]

File: src/dmd/astbase.d [hunk 1]
```diff
@@ -6056,9 +6067,9 @@ struct ASTBase
             return null;
         }
 
-        ExpInitializer isExpInitializer()
+        final ExpInitializer isExpInitializer()
         {
-            return null;
+            return kind == InitKind.exp ? cast(ExpInitializer)cast(void*)this : null;
```
Comment (Geod24) [inline src/dmd/astbase.d]: Since those are `C++` classes, aren't the `void*` cast useless ?

File: src/dmd/init.d [hunk 2]
```diff
@@ -36,15 +36,30 @@ enum NeedInterpret : int
 alias INITnointerpret = NeedInterpret.INITnointerpret;
 alias INITinterpret = NeedInterpret.INITinterpret;
 
+/*************
+ * Discriminant for which kind of initializer
+ */
+enum InitKind : ubyte
```
Comment (ibuclaw) [inline src/dmd/init.d]: Would this information be useful to C++ consumers? Or should they only be using visitors/the available isXxxInitializer() methods to get to the correct type?

Comment (WalterBright) [reply src/dmd/init.d]: I don't know. I'm not even sure why a C++ consumer would need init.h at all.

File: src/dmd/astbase.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/astbase.d]: You're right.

Comment (WalterBright) [reply src/dmd/astbase.d]: But it doesn't cause any problems, and if they are ever converted to D classes, adding the cast might be overlooked, and would be a hidden speed bump. I suppose I could argue it either way.

**Comment by WalterBright** [comment]
@ibuclaw Since `Initializer` is small, I'm using it to test out replacing the Visitor pattern with a switch statement. As you measured, it should give us a nice performance boost.

File: src/dmd/init.d (hunk 2 above)
Comment (ibuclaw) [reply src/dmd/init.d]: To generate code for the _init of a variable.

Having a quick/naive look (I have probably missed something), I can only see three points of interest.
- isVoidInitializer
- isExpInitializer
- initializerToExpression

Comment (WalterBright) [reply src/dmd/init.d]: If your code sticks with the `is` functions, I don't think you'll need the enum at all.

**Comment by jacob-carlborg** [comment]
I would have preferred going with a generic kind placed in either `RootObject` or a new base class which all other classes inherit.

**Comment by WalterBright** [comment]
> I could point you to one bug report

Sheesh. Just point to it!

**Comment by ibuclaw** [comment]
> Sheesh. Just point to it!

Oops, I thought i did.  Serves me right for replying via my phone.

https://issues.dlang.org/show_bug.cgi?id=6401

**Pull request merged**